### ISSUE TITLE
Plugin that detects files for which we do not have representation information (i.e. a format in the format registry)

### ISSUE THREAD

**Issue by jmaferreira** [keeps/roda#650]
Detect if the file format is identified on the File Format Registry. The plugin should:
- The plugin should operate over IndexedFiles
- The plugin should have 3 parameters flags (mime type, promo, format)
- The plugin should check for the existence of a specific Risk (if it doesn't exist, create it - see how "AIP corruption risk assessment" works).
- Affected files should be added as incidences to the created risk
- Make sure the plugin is well described
- Generate PREMIS event
- Put on the plugin outcome details the summary of checked files

**Comment by rui-castro** [comment]
https://github.com/rui-castro/roda/tree/rcastro_650_plugin_findunknownfiles/roda-core/roda-plugins/misc/roda-plugin-misc-findunknownfiles

**Comment by jmaferreira** [comment]
Looking at the source.. I don't think this does exactly what it was supposed to do. Lets talk personally.

**Comment by rui-castro** [comment]
@luis100 @hsilva-keep done in PR #703 

@jmaferreira The "format designation" is an attribute of the FileFormat associated with a File, but Format doesn't have a "format designation", right?
So, it's not possible to search for "format designations" in the Format Registry.
Therefore, this plugin has 2 parameters only (mimetype and pronom).

Am I missing something?

NOTE: the risk information (description and notes) needs to be improved. I don't know how...
https://github.com/rui-castro/roda/blob/rcastro_650_unknown_format_plugin/roda-core/roda-plugins/misc/roda-plugin-misc-unknownformat/src/main/resources/default/data/storage/risk/urn:unknownmimetype:r1.json
https://github.com/rui-castro/roda/blob/rcastro_650_unknown_format_plugin/roda-core/roda-plugins/misc/roda-plugin-misc-unknownformat/src/main/resources/default/data/storage/risk/urn:unknownpronom:r1.json

**Comment by jmaferreira** [comment]
Rui,

The plugin should have 4 parameters. All the selected attributes should match, otherwise, a risk should be created:

PREMIS - Format Registry
====================
Format name - name
Format version - latestVersion
Mimetype - mimetypes
PRONOM - pronoms

**Comment by rui-castro** [comment]
@jmaferreira currently the plugin is testing attributes (Mimetype and PRONOM) separatly. 

Like this (pseudo-code):
```
for file in files
  if not checkRegistryForFormatWithMimetype(file.mimetype): createRisk()
  if not checkRegistryForFormatWithPronom(file.pronom): createRisk()
```

But for `Format name` and `Format version` it only makes sense to test them together. Like this:
```
for file in files
  if not checkRegistryForFormatWithNameAndVersion(file.name, file.version): createRisk()
```

Should **all** attributes be tested at the same time? 
Something like this?
```
for file in files
  if not checkRegistryForFormat(file.name, file.version, file.mimetype, file.pronom): createRisk()
```

**Comment by luis100** [comment]
NPE happens when running the plugin:

```
2016-11-23 18:15:48,775 [JobsSystem-akka.actor.default-dispatcher-18] ERROR o.r.c.p.o.akka.AkkaWorkerActor - Error executing plugin.execute()
java.lang.NullPointerException: null
        at java.io.Reader.<init>(Reader.java:78)
        at java.io.InputStreamReader.<init>(InputStreamReader.java:113)
        at org.apache.commons.io.IOUtils.copy(IOUtils.java:2272)
        at org.apache.commons.io.IOUtils.toString(IOUtils.java:1041)
        at org.apache.commons.io.IOUtils.toString(IOUtils.java:1066)
        at org.roda.core.data.utils.JsonUtils.getObjectFromJson(JsonUtils.java:129)
        at org.roda.core.plugins.plugins.PluginHelper.createDefaultRisk(PluginHelper.java:480)
        at org.roda.core.plugins.plugins.PluginHelper.createRiskIfNotExists(PluginHelper.java:464)
        at org.roda.core.plugins.misc.UnknownFormatPlugin.createIncidence(UnknownFormatPlugin.java:323)
        at org.roda.core.plugins.misc.UnknownFormatPlugin.executeOnFile(UnknownFormatPlugin.java:252)
        at org.roda.core.plugins.misc.UnknownFormatPlugin.execute(UnknownFormatPlugin.java:125)
        at org.roda.core.plugins.orchestrate.akka.AkkaWorkerActor.handlePluginExecuteIsReady(AkkaWorkerActor.java:49)
        at org.roda.core.plugins.orchestrate.akka.AkkaWorkerActor.onReceive(AkkaWorkerActor.java:35)
        at akka.actor.UntypedActor$$anonfun$receive$1.applyOrElse(UntypedActor.scala:165)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
        at akka.actor.UntypedActor.aroundReceive(UntypedActor.scala:95)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
        at akka.actor.ActorCell.invoke(ActorCell.scala:495)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

**Comment by luis100** [comment]
Maybe due to the above failures, the plugin did not report correctly, ending with 95% done and several reports not getting out of processing state.

![screenshot de 2016-11-23 18-20-18](https://cloud.githubusercontent.com/assets/631728/20573822/8a747f98-b1a9-11e6-880b-c39668e06a1e.png)

**Comment by luis100** [comment]
The risk was not created and no incidences are to be found. Not clear if it is an error or if all my formats are known.

**Comment by luis100** [comment]
Note that code was already updated to set categories and improve some methods, please check branch unknown_and_missing_formats

**Comment by rui-castro** [comment]
@luis100 did you tested it on top of PR #701 ?
It looks like it's not finding a json that is in the plugin resources. That feature to load plugin resources is in PR #701.

**Comment by luis100** [comment]
Maybe the error is due to [one of my commits](https://github.com/keeps/roda/commit/518a0f3126a179a953a8fda5cbeedb04604fed70) during merge, will check it tomorrow.

**Comment by jmaferreira** [comment]
@rui-castro, regarding your question before... 

1. Yes, your are right.. format and version should be coupled together.
2. Regarding the behaviour of the plugin, I think the following sentence sums it all: All the attributes **selected** during the execution of the plugin (format name and version, PRONOM, MIME) should match the values found on the format registry, otherwise, a risk should be created. For example, if I only select "format name and version", the plugin will only look at this atribute set. If I select two options, both options must match the values on the format registry, otherwise, should create risk.

**Comment by rui-castro** [comment]
plugin class `org.roda.core.plugins.riskManagement.FormatMissingRepresentationInformationRiskAssessmentPlugin`

**Comment by rui-castro** [comment]
@luis100 @hsilva-keep @jmaferreira new requirements implemented in PR #716

**Comment by luis100** [comment]
- [ ] Output is unclear, does 1 means "yes" and 0 means "no"?

```
File is not comprehensively characterized.
Missing format designation: 1
Missing mimetype: 0
Missing PRONOM UID: 0
```

- [ ] Plugin created incidence when no incidence seems to be needed

Tried the plugin and it stated is was missing format format designation but it wasn't.

![report](https://cloud.githubusercontent.com/assets/631728/20680800/7eac436a-b598-11e6-840f-c3e8fd68bce1.png)
![file](https://cloud.githubusercontent.com/assets/631728/20680801/81216c38-b598-11e6-9b76-c4cc1f63dc0c.png)

**Comment by luis100** [comment]
```
File format has Representation information.
```
```
File format doesn't have Representation information.
```

Report information is too cryptic, not clear which representation information exists and where it is. Change to:
```
File format is described in format "Extensible Markup Language (XML) 1.0" (dd1e605a-fad6-422f-896c-85ca4057d0c8). Format description has MIME type and PRONOM but not the format designation.
```
Note: name and id are fetched from the format description hit.

```
Could not find a format description with MIME type "application/xml" or PRONOM "fmt/110" or format designation "Extensible Markup Language".
```
Note: it can be OR or AND depending on the options.

**Comment by rui-castro** [comment]
@luis100 I'm not sure your idea is possible...

Imagine you have a **File** with
**mimetype**: application/zip
**pronom**: fmt/666
**format designation name**: Quake pack file
**format designation version**: 1

In the format registry you have the following **Formats**:

| # | Name | Mimetype | Pronom |
|---|------|----------|--------|
|1|ZIP file|application/zip||
|2|Microsoft Office Word||fmt/666|
|3|Quake pack file, version 1|||

And the you run the plugin with all attributes selected (mimetype, pronom and format).

How would you know which one is the Format you should check?
You could say these 3 things
"File format is ZIP. Format has mimetype but it's missing pronom and format designation"
"File format is Word. Format has PRONOM but it's missing mimetype and format designation"
"File format is Quake pack. Format format designation, but it's missing mimetype and PRONOM.

And you still have the case where the File you're checking doesn't have the attributes you want to check.

**Comment by rui-castro** [comment]
@luis100 new commit with support to define the name pattern in roda properties.
The default name pattern is ```%1$s, version %2$s```.
It can be overwritten in roda-core.properties using property ```core.plugins.external.FormatMissingRepresentationInformation.format_name_pattern```

This name pattern is used to combine the format designation name and version into the Format name. The plugin searches for this name in the Format registry.

A file with 
formatDesignationName "Acrobat PDF 1.5 - Portable Document Format" and 
formatDesignationVersion "1.5" has 
format named "**Acrobat PDF 1.5 - Portable Document Format**, version **1.5**".

**Comment by luis100** [comment]
1. If file doesn't have attribute, create report failure stating that "File does not have required information (format designation, MIME type or PRONOM), to be able to find format representation information". Don't create incidence.
2. If file has attributes but cannot find any format that fits the requirements, create report failure stating that and create incidence.
3. If file has attributes and can find one or more formats that fits the requirements, create report success and list all formats that fit identifying the information each one has, e.g. has format designation, does NOT have format designation version, does NOT have MIME type, has PRONOM.

**Comment by rui-castro** [comment]
@luis100 @hsilva-keep @jmaferreira implemented in PR #716 (this is the third implementation of this plugin).

**Comment by luis100** [comment]
```
Format "Extensible Markup Language (XML) 1.0"
	doest NOT have Format designation "Extensible Markup Language, version 1.0"
	has Mimetype "application/xml"
	has PRONOM "fmt/101"
```
- [ ] Missing format id: Format "Extensible Markup Language (XML) 1.0" (dd1e605a-fad6-422f-896c-85ca4057d0c8)
- [ ] typo: "does**t** NOT ..."

**Comment by luis100** [comment]
Merged into lf_dev at 143724e6967e5c1b8c07c81ef23b4a67ccd10872 but waiting for the last comment feedback to close issue.

**Comment by rui-castro** [comment]
@luis100 done in PR #737

**Comment by rui-castro** [comment]
@luis100 this can be closed, right?

**Comment by luis100** [comment]
Merged into lf_dev at 8d94866eed351c7942d8c898cae41232ed33b7be
